src/main/java/com/atlassian/db/replica/spi/ConnectionPoolPerUrlProvider.java

```diff
@@ -0,0 +1,6 @@
+package com.atlassian.db.replica.spi;
+
+@FunctionalInterface
+public interface ConnectionPoolPerUrlProvider {
+    ConnectionPool getConnectionPool(String url);
```

Comment:
Suppose it's per URL then java.net.URL instead of String. Or maybe it's not (doesn't have to be) URL, but id. String id would be more generic.

Reply:
- IMO introducing java.net.URL is unnecessary complexity
- url clearly indicates what has to be passed into the method; id might be confusing since node id is not necessarily equal to node url

I could rename to `-PerReplica-` instead of `-PerUrl-` if that would increase the readability?

Reply:
> IMO introducing java.net.URL is unnecessary complexity

We have a type to represent URI build in the language. Why increasing type safety is more complex? Using Objects instead of string would be even less complex?

It's a string, so any string can be passed. 

> id might be confusing since node id is not necessarily equal to node URL

It doesn't have to be node id.

Reply:
- introducing java.net.URL will require us to handle all the exceptions (simplest case scenario we just rethrow them or do not catch them at all).
- jdbc drivers already handle the url string when creating database connections. handling the url on our side would be just duplicate work
- jdbc drivers use different protocols than java.net.URL expects. e.g. java.net.URL does not recognise the "jdbc:postgresql" protocol and raises a `java.net.MalformedURLException` while trying to instantiate the class. Can you provide a way to transform a database String url into a java.net.URL?

Reply:
> introducing java.net.URL will require us to handle all the exceptions (simplest case scenario we just rethrow them or do not catch them at all).

URI/URL have a specified syntax, and the type ensures it's what it should be. It gives us safety for free (fail fast). Other options are to validate the syntax on our own on pass it on. It may be harder to debug and understand why something is failing.

> jdbc drivers already handle the url string when creating database connections.

JDBC is a very stable API. It will contain lots of legacy stuff. But we don't need to inherit it. We can be more type-safe than JDBC. 

> jdbc drivers use different protocols than java.net.URL

Use URI.

Reply:
> URI/URL have a specified syntax, and the type ensures it's what it should be

But the valid syntax for a database url is only a subset of what is permissible by the URI class. e.g: `http://localhost:1234/index.html` is a valid URI syntax, but it is not a valid database url syntax. 

> It gives us safety for free (fail fast)

URI does not guarantee that the underneath string is a valid database url, so this is not fail-fast. 
Neither is it free because we would be paying the overhead of parsing a string into a URI.

If we want to pass a database url in a type-safe manner, then we should write our own class that checks for the validity of a database url.

Reply:
Yes. URI is more type-safe than String but could be more strict if we introduce our own type.

Reply:
would `AuroraJdbcUrl` be a good candidate for this?

Reply:
obviously, it would have to be in the api package and be aurora-agnostic but it's just a suggestion

Reply:
> JdbcUrl) and Aurora-specific implementation (URI is also more generic).

We can also use URI with the partial type safety and let the implementation assert on DB specific aspects.

Reply:
Introduced a JdbcUrl class that's a wrapper for a database url string.
Added a Builder class for manually constructing a valid database url.
Re-used existing validity checks of the `java.net.URI` class